aws/graph-explorer
# Move node count logic to one place

## Description



- Updates node counts when new nodes are added to the graph
- Adds caching and retries to node expansion and neighbor counts
- Use React Query for node expansion and neighbor counts
- Add notifications for progress of node expansion and neighbor counts
- Add notifications for errors during node expansion and neighbor counts
- Remove "limit neighbors" option
- Adds connection details to `Explorer` instance so the explorers will be not equal by value

This change pulls all queries related to getting the neighbor counts to a single place. It watches for changes to the list of nodes in the graph. When a node is added we execute a query for that nodeId.

Neighbor expansion is also a React Query now. It is hosted in a React context to share the expand request state. This ensures a single expand is allowed to execute at any one time. And it also allows better control over notifications to the user so there is only one error notification and one progress notification. I would've used Recoil state instead, but I ran in to problems with that approach.

I also removed the limit parameter when executing the callback from `useFetchNode()`. This limit was fairly pointless since it only affected that one node, not any future nodes. And the behavior was strange. In search it would limit to 500 when the check box is checked. In the DataExplorer it will limit the added node to the page size.

## Validation



- Tested in all query languages
- Manually threw error in explorer to simulate error responses from the server
- Checked caching across connections and reset logic by:
  1. Add node to graph
  2. Expand node
  3. Change query engine for same database
  4. Add same node and expand
  5. Observe network requests

## Related Issues


- Resolves #338 
- Resolves #339 

### Check List



      license.
